File: q2.java
import java.util.*;

class TestClass {

    // Inner class to encapsulate the logic for finding Articulation Points (Type A cities) 
    // and Type B cities.
    static class Solution {
        private int N;
        private int[][] adj;
        private int[] disc; // Discovery time
        private int[] low;  // Low-link value
        private int[] parent;
        private int time;
        private Set<Integer> articulationPoints; // Set of Type A cities

        public Solution(int[][] arr, int n) {
            this.N = n;
            this.adj = arr;
            this.disc = new int[N];
            this.low = new int[N];
            this.parent = new int[N];
            this.time = 0;
            this.articulationPoints = new HashSet<>();

            // Initialize arrays for DFS
            Arrays.fill(disc, -1);
            Arrays.fill(parent, -1);
        }

        /**
         * Finds all articulation points (Type A cities) in the graph.
         */
        public void findArticulationPoints() {
            // Iterate through all vertices in case the graph is disconnected
            for (int i = 0; i < N; i++) {
                if (disc[i] == -1) {
                    DFS(i);
                }
            }
        }

        /**
         * DFS traversal to calculate discovery times and low-link values, 
         * and identify articulation points.
         */
        private void DFS(int u) {
            int children = 0;
            // Initialize discovery time and low-link value for u
            disc[u] = low[u] = ++time;

            // Iterate through neighbors of u
            for (int v = 0; v < N; v++) {
                // Check if v is a neighbor (adj[u][v] == 1) and not the same vertex (u != v)
                if (adj[u][v] == 1 && u != v) {
                    
                    if (disc[v] == -1) { // v is not visited (Tree edge)
                        children++;
                        parent[v] = u;
                        DFS(v);

                        // Update low[u] based on low[v]
                        low[u] = Math.min(low[u], low[v]);

                        // Check if u is an articulation point based on DFS criteria:
                        
                        // 1. If u is the root of the DFS tree and has at least two children.
                        if (parent[u] == -1 && children > 1) {
                            articulationPoints.add(u);
                        }
                        
                        // 2. If u is not the root, and there is a child v such that low[v] >= disc[u].
                        if (parent[u] != -1 && low[v] >= disc[u]) {
                            articulationPoints.add(u);
                        }
                    } else if (v != parent[u]) { // v is visited and not the parent (Back edge)
                        // Update low[u] using the discovery time of v
                        low[u] = Math.min(low[u], disc[v]);
                    }
                }
            }
        }

        /**
         * Identifies Type B cities.
         * Type B cities are defined as those connected directly to Type A cities.
         * * @return A sorted set of Type B city indices.
         */
        public Set<Integer> findTypeBCities() {
            // Ensure articulation points are identified first
            findArticulationPoints();

            // Use TreeSet to store Type B cities, which automatically keeps them sorted.
            Set<Integer> typeBCities = new TreeSet<>();

            // Iterate through all cities
            for (int city = 0; city < N; city++) {
                
                // A city cannot be both Type A and Type B simultaneously based on the definition
                if (articulationPoints.contains(city)) {
                    continue;
                }

                // Check if the current city is directly connected to any Type A city
                for (int ap : articulationPoints) {
                    if (adj[city][ap] == 1) {
                        typeBCities.add(city);
                        // Once connected to one Type A city, it's a Type B city.
                        break; 
                    }
                }
            }
            return typeBCities;
        }
    }

    // The main function to solve the problem
    static void connectedVertex(int[][] arr, int n) {
        // Instantiate the solution helper with the provided graph data
        Solution solver = new Solution(arr, n);
        
        // Find Type B cities
        Set<Integer> typeBCities = solver.findTypeBCities();
        
        // Print Type B cities in ascending order separated by single space
        StringBuilder result = new StringBuilder();
        for (int city : typeBCities) {
            result.append(city).append(" ");
        }
        
        // Print the result, trimming the trailing space
        if (result.length() > 0) {
            System.out.println(result.toString().trim());
        }
    }

    // Main method provided in the template for input handling.
    public static void main(String[] args) {
        // We use Scanner for input as specified in the provided Java code snippet (File 3).
        Scanner input = new Scanner(System.in);
        
        // Read N (number of cities)
        int n = input.nextInt();
        
        // Read the N x N adjacency matrix
        int arr[][] = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                arr[i][j] = input.nextInt();
            }
        }
        
        // Call the method to find and print Type B cities
        connectedVertex(arr, n);
        
        input.close();
    }
}

File: q2.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <utility> // for std::min

using namespace std;

// Helper function for DFS-based Articulation Point (Type A) finding.
// This function relies on external state passed by reference for discovery times, low-link values, 
// parent tracking, a timer, and the set of articulation points.
void find_aps_dfs(int u, int N, int** adj, vector<int>& disc, vector<int>& low, 
                  vector<int>& parent, int& timer, set<int>& aps) {
    
    int children = 0;
    disc[u] = low[u] = ++timer;

    // Iterate through all cities to find neighbors (using adjacency matrix)
    for (int v = 0; v < N; ++v) {
        if (adj[u][v] == 1 && u != v) {
            if (disc[v] == -1) { // Tree edge
                children++;
                parent[v] = u;
                
                // Recurse
                find_aps_dfs(v, N, adj, disc, low, parent, timer, aps);

                // Update low[u] based on low[v]
                low[u] = min(low[u], low[v]);

                // Check for Articulation Point criteria:
                // 1. Root of DFS tree with >= 2 children
                if (parent[u] == -1 && children > 1) {
                    aps.insert(u);
                }
                // 2. Non-root, and no back edge from v's subtree to an ancestor of u
                if (parent[u] != -1 && low[v] >= disc[u]) {
                    aps.insert(u);
                }
            } else if (v != parent[u]) { // Back edge
                low[u] = min(low[u], disc[v]);
            }
        }
    }
}

// Function that identifies and prints Type B cities following the template.
void connectedVertex(int **arr, int n) {
    // WRITE YOU CODE HERE

    // 1. Setup for Articulation Points (Type A) finding
    vector<int> disc(n, -1), low(n, -1), parent(n, -1);
    int timer = 0;
    set<int> typeA; // Type A cities (Articulation Points)

    // Run DFS to find all articulation points
    for (int i = 0; i < n; ++i) {
        if (disc[i] == -1) {
            find_aps_dfs(i, n, arr, disc, low, parent, timer, typeA);
        }
    }

    // 2. Identify Type B cities
    // Type B cities are connected directly to Type A cities, but are not Type A themselves.
    set<int> typeB; // Use set for sorted output

    for (int i = 0; i < n; ++i) {
        if (typeA.count(i)) continue; // Skip if Type A

        for (int ap : typeA) {
            // Check connectivity between city i and articulation point ap
            if (arr[i][ap] == 1) {
                typeB.insert(i);
                break; // Found connection, move to the next city
            }
        }
    }

    // 3. Print Type B cities in ascending order separated by a single whitespace.
    // // Print Type B cities in ascending order separated by a single whitespace.
    bool first = true;
    for (int city : typeB) {
        if (!first) cout << " ";
        cout << city;
        first = false;
    }
}

// The main function structure provided in the template:
// int main() {
//     int n;
//     cin >> n;
//     int **arr;
//     arr = new int*[n];
//     for(int i=0; i<n; ++i){
//         arr[i] = new int[n];
//         for(int j=0; j<n; ++j){
//             cin >> arr[i][j];
//         }
//     }
//     
//     connectedVertex(arr, n);
//     return 0;
// }
